### PR TITLE
Don't save device selection when there is only one device

### PR DESCRIPTION
**Pull request by nickvergessen**
This will allow a laptop to recover after being disconnected from
a docking-station for one call and being docked afterwards again
to use the previous selected device again.

### CODE REVIEW

File: src/utils/webrtc/MediaDevicesManager.js [hunk 1]
```diff
@@ -126,10 +126,23 @@ MediaDevicesManager.prototype = {
...
+			} else {
+				BrowserStorage.removeItem(key)
+			}
 		} else {
-			BrowserStorage.removeItem(key)
+			// Not saving device selection as there is only one device.
+			// This will allow a laptop to recover after being disconnected from
+			// a docking-station for one call and being docked afterwards again
+			// to use the previous selected device again.
+			console.debug('Not saving device selection as there is only one device')
```
Comment (marcoambrosini) [inline src/utils/webrtc/MediaDevicesManager.js:145]: Why the need for logging this?

**Review by marcoambrosini — APPROVED** [review]

**Review by danxuliu — CHANGES_REQUESTED** [review]
The devices are only set from MediaDevicesPreview (and from MediaDevicesManager, but if the user manually overrode the device from the media permissions dialog, which was only possible in Firefox and, if I am not mistaken, it is no longer possible either due to some changes in the constraints used now). The MediaDevicesManager just stores whatever it is told to store, so this should be addressed in the MediaDevicesPreview component instead.

As far as I can tell the problem is that as soon as the MediaDevicesPreview is opened it stores the selected device, even if the user has not explicitly selected it. Therefore, if the user opens the settings, for example to check the shortcuts, whatever audio and video devices are currently the default ones will be stored, even if the audio and video devices were not explicitly selected.

If the MediaDevicesPreview is not opened everything works as expected (without the changes in this pull request). If you select a device and start a call the selected device is used. Leave the call, disconnect the device and start a call the default device will be used. If you then leave the call, connect the previous device and join the call then the previous device will be selected again.

Thus, what needs to be fixed is that the MediaDevicesPreview only calls `mediaDevicesManager.set('audio/videoInputId', XXX)` when the user explicitly selects an item from the dropdown.

File: src/utils/webrtc/MediaDevicesManager.js [hunk 2]
```diff
@@ -126,10 +128,23 @@ MediaDevicesManager.prototype = {
 			value = LOCAL_STORAGE_NULL_DEVICE_ID
 		}
 
-		if (value) {
-			BrowserStorage.setItem(key, value)
+		const numberOfDevices = this.attributes.devices.filter(device => {
```
Comment (danxuliu) [inline src/utils/webrtc/MediaDevicesManager.js:131]: I would prefer all this to be done in the `MediaDevicesPreview` methods using the new `storeInBrowser` parameter.

Comment (nickvergessen) [reply src/utils/webrtc/MediaDevicesManager.js:131]: But we also need it with `_updateSelectedDevicesFromGetUserMediaResult` or we make that never store (would also be fine for me)

**Review by danxuliu — COMMENTED** [review]
When using Chromium the device is always stored, because Chromium adds a virtual "default" device so there are more than a real device and a "None" option.

But... if you use a virtual camera then there is no "default" device, and yet it also stores the camera even if in the dropdown there is only a camera and "None". ~~I do not know why :shrug:~~ Ah, ok, it is because `this.devices` is not filtered by kind, so it contains all audio and video devices.

File: src/utils/webrtc/MediaDevicesManager.js (hunk 2 above)
Comment (danxuliu) [reply src/utils/webrtc/MediaDevicesManager.js]: This should not be a problem, as `_updateSelectedDevicesFromGetUserMediaResult` is most likely dead code now since the fix for switching devices in Firefox. But I have not checked, so better keep it just in case ;-) Anyway, not storing there should not be a problem.

**Comment by nickvergessen** [comment]
> this.devices is not filtered by kind, so it contains all audio and video devices. 

happened to be 2 with fake stream, so I thought it's it, but yeah was fake camera and mic instead of fake + none for each.

**Review by danxuliu — COMMENTED** [review]
- Open Talk on Chromium with just one microphone (so the list contains "Default", the microphone and none)
- Open the settings; the default device will not be selected
- Select "None"; the null id is written to the store
- Select "Default"; default is not written to the store, you need to select the microphone instead

**Comment by nickvergessen** [comment]
Arg, so yeah I found that, but it actually means it is breaking the feature on chromium as the default will mean there are always 2 devices.
Will think a bit about it and see if we can filter that out or something.

**Review by danxuliu — COMMENTED** [review]
Filtering by `device.id !== 'default'` does not work, because the filtered array is `this.devices`, so it should be `device.deviceId !== 'default'` instead (id is only for options).

However, although ignoring the default device seems reasonable it also means that if the user switches between "Default" and the actual device that will not be stored, and when Talk is opened again the previous option will be selected. I am not sure if that could be a surprising behaviour for the user, even if in practice the device is the same :shrug:

**Review by danxuliu — APPROVED** [review]
When permissions were not granted yet when the settings were opened the options were loaded with the fallback names. Once the permissions were granted the options were updated with the real labels. This caused the `deviceSelectedOption` to change, even if the id was still the same, which caused the event to be triggered. I have added a fix to ignore that case.

Another case in which the device was wrongly updated was when the settings were open and the current device was disconnected. This updates the selected options and caused the fallback to be stored as the one selected by the user. I have added a fix for that too.

Turns out that, with those two fixes and the fix to ignore the change when the settings were initially loaded, the code to avoid storing the selected device when there is only one device is no longer needed. If there is only one device, the dropdown is opened and the device is selected again the option is not changed, so `deviceSelectedOption` does not change and no event is triggered (so if the user removes the camera and opens again the settings the fallback will be selected, and clicking on it will not store it, so if the camera is connected again it will be automatically selected). An event is only triggered if the device is actually changed by the user (for example, to set "None", and from "None" to the device, but this case had to be taken into account already in the other code).

The only difference is that, without that explicit code, if you are using Chromium and you switch between "Default" and the device now when there is only one device that triggers an event and the device is stored in the browser, but as mentioned already not doing that could be a surprising behaviour for the user (and if the user does not switch the device but clicks again on the same option used before then no event is triggered).

Regarding `_updateSelectedDevicesFromGetUserMediaResult` now the selected device is stored again, but this is probably dead code now and, even if it is not, it would reflect what the user selected in the browser UI. I would be fine not storing the device in that case, but it would be a change most likely with no effect.

Nevertheless, I split the fixes in three fixup commits. The first one to handle granting media permissions. The second one to not store the fallback device when the current device is disconnected. And the third one to remove the special handling when there is only one device. The two first fixups are needed; the third one is, from my point of view, a better approach, and based on my tests it should fix the original issue.

To sum up, tested (with Firefox and Chromium, one device and several devices, connecting and disconnecting them, reopening the browser, before and during a call) and, as far as I can tell, works :+1:

File: src/components/MediaDevicesSelector.vue [hunk 3]
```diff
@@ -155,7 +155,26 @@ export default {
...
+			// could be triggered if media permissions are granted, which would
+			// update the label.
+			if (deviceSelectedOption && previousSelectedOption && deviceSelectedOption.id === previousSelectedOption.id) {
+				return
+			}
+
+			// The previous selected option changed due to the device being
+			// disconnected, so ignore it as it was not explicitly changed by
+			// the user.
+			if (previousSelectedOption && previousSelectedOption.id && !this.deviceOptions.find(option => option.id === previousSelectedOption.id)) {
```
Comment (nickvergessen) [inline src/components/MediaDevicesSelector.vue:169]:
```suggestion
			if (previousSelectedOption?.id && !this.deviceOptions.find(option => option.id === previousSelectedOption.id)) {
```

**Comment by nickvergessen** [comment]
Backported via https://github.com/nextcloud/spreed/pull/6179
